### DISCUSSION TITLE
How to use @kind graph query?

### DISCUSSION THREAD

**Discussion by ox1234** [github/codeql#7437] [Q&A] [answered]
I use @kind graph query and got such error:
![image](https://user-images.githubusercontent.com/32171650/146515828-760f9578-0f54-447b-9262-9ce0d7cdb028.png)
Here is my select:
![image](https://user-images.githubusercontent.com/32171650/146515998-4ba817f6-c05b-473f-a833-9b9121d2f24d.png)

**Comment by intrigus-lgtm** [comment, marked as answer]
The query as you have written requires `@kind path-problem` and not `@kind graph`.

Documentation on the usage of `@kind graph` is hard to find, but it seems to work like this:
You have `@kind graph` as metadata and two `query predicate`s. (You don't define any `from`, `where`, or `select` clause)
The first predicate defines the nodes of the graph and the second defines the edges.
The prototype of `nodes` is `query predicate nodes(ANY_TYPE node, string attr, string val)` and the prototype of `edges` is `query predicate edges(ANY_TYPE pred, ANY_TYPE succ, string attr, string val)`
See here for some kind of example:
https://github.com/github/codeql/blob/15caaa7ad6d91c533596204b6a863a1c56240628/ruby/ql/test/library-tests/regexp/parse.ql#L8-L27

**Reply by ox1234** [reply]
Thanks for a lot, but i'm still don't understand the meaning of attr and val.

**Reply by ox1234** [reply]
Do i need to define all showed nodes in nodes predicate?
I write following ql to draw the call graph from the entry, but i don't get any edges in generated .dot file
```codeql
query predicate nodes(InterestringCallable node, string attr, string val){
    exists(node) and attr = "semmle.label" and val = getCallableSig(node)
}

query predicate edges(Callable pred, Callable succ, string attr, string val){
    succ = pred.getACallee() and
    attr in ["semmle.label"] and
    val = getCallableSig(succ)
}
```

**Reply by ox1234** [reply]
I draw a graph successfully, thanks a lot for your help😊
